File: cmd/razvilka/web/panel-snapshot.js
'use strict';

// This fallback contains saved choices, never a route recommendation or proof.
function validSavedPanel(value) {
  return value?.schema === 1 && value.dataplane === 'not-checked'
    && ['empty', 'available', 'retained', 'expired'].includes(value.state)
    && Number.isSafeInteger(value.revision) && value.revision >= 0
    && Number.isFinite(value.data_age_ms) && value.data_age_ms >= 0
    && Number.isFinite(value.max_age_seconds) && value.max_age_seconds > 0
    && value.max_age_seconds <= 900
    && (!value.data || (typeof value.instance_id === 'string' && value.instance_id.length === 32
      && Number.isSafeInteger(value.data.config?.revision)
      && Array.isArray(value.data.services) && value.data.services.length <= 512
      && value.data.services.every(s => typeof s.id === 'string' && typeof s.name === 'string'
        && typeof s.desired?.enabled === 'boolean' && typeof s.applied?.enabled === 'boolean'
        && typeof s.desired.route === 'string' && typeof s.applied.route === 'string'
        && (s.sources == null || Array.isArray(s.sources) && s.sources.every(x => typeof x === 'string'))
        && (s.applied_sources == null || Array.isArray(s.applied_sources) && s.applied_sources.every(x => typeof x === 'string')))));
}

async function loadSavedPanelSnapshot(generation, options) {
  try {
    const value = await api('/api/v1/panel/snapshot', { ...options, readTimeoutMs: 3000 });
    if (!panelSnapshotCurrent(generation) || !state.authenticated || !validSavedPanel(value)) return;
    const previous = state.savedPanel?.value;
    if (previous?.instance_id === value.instance_id && previous.revision > value.revision) return;
    state.savedPanel = { value, receivedAt: Date.now() };
    if (state.dataLoad?.metrics?.phase !== 'ready' && Number.isFinite(Date.parse(value.metrics?.timestamp))) {
      state.metrics = { ...(state.metrics || {}), latest: value.metrics };
      state.dataLoad.metrics = { loaded: true, phase: 'busy', message: 'Последнее измерение роутера' };
      try { renderMetrics(); } catch (_) { /* A chart cannot gate navigation. */ }
    }
    renderPanelLoad();
  } catch (error) {
    if (!panelSnapshotCurrent(generation)) return;
    if (error.status === 401) {
      cancelPanelRefresh();
      showAuth({ authenticated: false }, 'Сессия завершилась. Войдите снова.');
    }
    // No replacement with empty configuration after a timeout or mixed version.
  }
}

function renderSavedPanelSnapshot() {
  const saved = state.savedPanel;
  const banner = $('#panelSavedNotice');
  if (!state.authenticated || !saved) { if (banner) banner.hidden = true; return; }
  const value = saved.value;
  const age = value.data_age_ms + Math.max(0, Date.now() - saved.receivedAt);
  const usable = value.data && age < value.max_age_seconds * 1000
    && ['available', 'retained'].includes(value.state);
  const load = state.dataLoad?.services;
  const mayReplace = !load?.loaded || ['busy', 'error'].includes(load.phase);
  if (usable && mayReplace && (load?.savedRevision !== value.revision || load?.savedInstance !== value.instance_id)) {
    state.services = value.data.services.map(s => ({
      id: s.id, name: s.name, category: s.category, custom: s.custom,
      enabled: s.desired.enabled, applied_enabled: s.applied.enabled,
      route: s.desired.route, mode: s.desired.route, applied_route: s.applied.route,
      sources: s.sources || [], applied_sources: s.applied_sources || [],
      desired_state: s.desired, applied_state: s.applied,
      dirty: s.route_dirty || s.sources_dirty, route_dirty: s.route_dirty, sources_dirty: s.sources_dirty,
      evidence_level: 'none', evidence_status: 'unknown', presentation_only: true,
    }));
    state.dataLoad.services = { loaded: true, phase: 'busy', savedRevision: value.revision,
      savedInstance: value.instance_id, updatedAt: Date.parse(value.generated_at),
      message: 'Показаны сохранённые настройки. Текущее подключение ещё проверяется.' };
  } else if (!usable && (load?.savedInstance || state.services?.some(s => s.presentation_only))) {
    // An expired image is unavailable, not evidence of removed services.
    state.services = [];
    state.dataLoad.services = { loaded: false, phase: 'error', message: 'Сохранённые данные устарели. Обновите состояние.' };
  }
  const showing = !!state.dataLoad?.services?.savedInstance;
  if (banner) {
    banner.hidden = !showing && (usable || state.dataLoad?.services?.loaded === true);
    banner.textContent = showing
      ? `Сохранённые настройки · ${Math.floor(age / 1000)} сек назад. Просмотр доступен; управление вернётся после обновления состояния. Это не результат проверки подключения.`
      : !usable ? 'Свежий снимок настроек пока недоступен. Это не означает, что сервисы удалены.' : '';
  }
}

document.addEventListener('razvilka:auth-required', () => {
  state.savedPanel = null;
  state.inventoryObservation = null; state.inventoryInvalidatedAt = 0;
  state.inventoryUnsupported = false;
  if (typeof cancelPanelInventoryRefresh === 'function') cancelPanelInventoryRefresh();
  state.components = []; state.engines = []; state.engineConfigs = []; state.system = {};
  // Clear protected presentation and freshness markers before the next login.
  state.services = []; state.sources = []; state.devices = [];
  state.nodes = { available: false, nodes: [], sources: [], counts: {} };
  state.dataLoad = {};
  const banner = $('#panelSavedNotice');
  if (banner) { banner.hidden = true; banner.textContent = ''; }
  for (const id of ['ui3ServiceCatalog', 'ui3HomeServices', 'serviceList', 'ui3InspectorContent']) {
    const element = document.getElementById(id);
    if (element) element.textContent = '';
  }
});

File: internal/app/panel_snapshot.go
package app

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"net/http"
	"slices"
	"sort"
	"sync"
	"sync/atomic"
	"time"

	"github.com/ArtixSx/razvilka/internal/catalog"
	"github.com/ArtixSx/razvilka/internal/operationgate"
	"github.com/ArtixSx/razvilka/internal/routerstats"
)

const panelSnapshotPath = "/api/v1/panel/snapshot"
const panelSnapshotLifetime = 15 * time.Minute
const maxPanelSnapshotBytes = 256 << 10

// These are presentation DTOs, never authority for an Apply or a health proof.
// In particular an applied AUTO choice is not the resolved/observed route.
type panelSavedService struct {
	ID             string                `json:"id"`
	Name           string                `json:"name"`
	Category       string                `json:"category"`
	Custom         bool                  `json:"custom"`
	Desired        serviceRouteStateView `json:"desired"`
	Applied        serviceRouteStateView `json:"applied"`
	Sources        []string              `json:"sources"`
	AppliedSources []string              `json:"applied_sources"`
	RouteDirty     bool                  `json:"route_dirty"`
	SourcesDirty   bool                  `json:"sources_dirty"`
}

type panelSavedConfig struct {
	Revision        uint64 `json:"revision"`
	AppliedRevision uint64 `json:"applied_revision"`
	SafeMode        bool   `json:"safe_mode"`
	Stopped         bool   `json:"stopped"`
	Mode            string `json:"mode"`
}

type panelSavedSource struct {
	ID             string    `json:"id"`
	Name           string    `json:"name"`
	Kind           string    `json:"kind"`
	Enabled        bool      `json:"enabled"`
	AppliedEnabled bool      `json:"applied_enabled"`
	CacheStatus    string    `json:"cache_status"`
	UpdatedAt      time.Time `json:"updated_at"`
}

type panelSavedJob struct {
	ID        uint64 `json:"id"`
	Kind      string `json:"kind"`
	State     string `json:"state"`
	Phase     string `json:"phase"`
	Total     int    `json:"total"`
	Completed int    `json:"completed"`
}

type panelSavedData struct {
	Config   panelSavedConfig    `json:"config"`
	Services []panelSavedService `json:"services"`
	Sources  []panelSavedSource  `json:"sources"`
}

// The immutable encoded document is detached from all Store-owned slices.
// The HTTP path only loads this pointer; it never acquires operation admission.
type panelSavedPublication struct {
	InstanceID string
	Revision   uint64
	Generated  time.Time
	Data       json.RawMessage
}

type panelSnapshotState struct {
	once    sync.Once
	latest  atomic.Pointer[panelSavedPublication]
	wake    chan struct{}
	done    chan struct{}
	started bool // set before serving HTTP; lifecycle methods must not race startup
}

type panelSnapshotResponse struct {
	Schema        int                    `json:"schema"`
	InstanceID    string                 `json:"instance_id"`
	Revision      uint64                 `json:"revision"`
	GeneratedAt   *time.Time             `json:"generated_at"`
	ObservedAt    time.Time              `json:"observed_at"`
	DataAgeMS     int64                  `json:"data_age_ms"`
	MaxAgeSeconds int64                  `json:"max_age_seconds"`
	State         string                 `json:"state"`
	Dataplane     string                 `json:"dataplane"`
	Admission     operationgate.Snapshot `json:"admission"`
	Data          json.RawMessage        `json:"data,omitempty"`
	Job           *panelSavedJob         `json:"job"`
	Metrics       routerstats.Snapshot   `json:"metrics"`
}

// StartPanelSnapshots must run after boot Store/journal recovery, before HTTP.
// A single publisher captures memory-only settings under a short exclusive
// lease. It never waits for a worker, downloads lists, probes engines, resolves
// DNS or reads a transaction journal. Busy/recovery leaves the old image intact.
func (a *App) StartPanelSnapshots(ctx context.Context) {
	a.panelSnapshots.once.Do(func() {
		p := &a.panelSnapshots
		p.wake, p.done, p.started = make(chan struct{}, 1), make(chan struct{}), true
		var id [16]byte
		if _, err := rand.Read(id[:]); err != nil {
			close(p.done)
			return // An unavailable presentation must not prevent recovery/Stop.
		}
		instance := hex.EncodeToString(id[:])
		a.publishPanelSnapshot(ctx, instance, time.Now())
		go func() {
			defer close(p.done)
			ticker := time.NewTicker(2 * time.Second)
			defer ticker.Stop()
			for {
				select {
				case <-ctx.Done():
					return
				case <-ticker.C:
				case <-p.wake:
				}
				a.publishPanelSnapshot(ctx, instance, time.Now())
			}
		}()
	})
}

func (a *App) wakePanelSnapshot() {
	if !a.panelSnapshots.started {
		return
	}
	select {
	case a.panelSnapshots.wake <- struct{}{}:
	default:
	}
}

func (a *App) WaitPanelSnapshots(ctx context.Context) error {
	if !a.panelSnapshots.started {
		return nil
	}
	select {
	case <-a.panelSnapshots.done:
		return nil
	case <-ctx.Done():
		return ctx.Err()
	}
}

func (a *App) publishPanelSnapshot(ctx context.Context, instance string, now time.Time) bool {
	release, err := a.Operations.Exclusive(ctx)
	if err != nil {
		return false
	}
	defer release()
	if a.Store == nil {
		return false
	}
	cfg := a.Store.Get()
	data := panelSavedData{Config: panelSavedConfig{Revision: cfg.Revision, AppliedRevision: cfg.AppliedRevision, SafeMode: cfg.SafeMode, Stopped: cfg.ServiceControl.Stopped, Mode: cfg.ServiceControl.EffectiveMode()}, Services: []panelSavedService{}, Sources: []panelSavedSource{}}
	// Do not use catalogSnapshot: source enrichment reads list files. Only
	// metadata is needed for navigation; definitions/proofs require a live read.
	definitions := append([]catalog.Service{}, a.Catalog.Services...)
	custom := map[string]bool{}
	if a.CustomServices != nil {
		for _, service := range a.CustomServices.List() {
			definitions = append(definitions, service)
			custom[service.ID] = true
		}
	}
	if len(definitions) > 512 {
		return false
	}
	for _, service := range definitions {
		desired, applied := cfg.Services[service.ID], cfg.AppliedServices[service.ID]
		data.Services = append(data.Services, panelSavedService{
			ID: service.ID, Name: service.Name, Category: service.Category, Custom: custom[service.ID],
			Desired: serviceRouteStateView{Enabled: desired.Enabled, Route: selectedRoute(desired), Source: "saved-settings"},
			Applied: serviceRouteStateView{Enabled: applied.Enabled, Route: selectedRoute(applied), Source: "saved-settings"},
			Sources: slices.Clone(desired.Sources), AppliedSources: slices.Clone(applied.Sources),
			RouteDirty:   desired.Enabled != applied.Enabled || selectedRoute(desired) != selectedRoute(applied),
			SourcesDirty: !stringSlicesEqual(desired.Sources, applied.Sources),
		})
	}
	sort.Slice(data.Services, func(i, j int) bool { return data.Services[i].ID < data.Services[j].ID })
	if a.Sources != nil {
		for _, source := range a.Sources.List() { // List is memory-only; URLs/errors are intentionally excluded.
			data.Sources = append(data.Sources, panelSavedSource{source.ID, source.Name, source.Kind, source.Enabled, source.AppliedEnabled, source.CacheStatus, source.UpdatedAt})
		}
	}
	encoded, err := json.Marshal(data)
	if err != nil || len(encoded) > maxPanelSnapshotBytes || ctx.Err() != nil {
		return false
	}
	revision := uint64(1)
	if previous := a.panelSnapshots.latest.Load(); previous != nil {
		revision = previous.Revision + 1
	}
	a.panelSnapshots.latest.Store(&panelSavedPublication{instance, revision, now.UTC(), encoded})
	return true
}

func (a *App) panelSnapshotAt(now time.Time) panelSnapshotResponse {
	value := panelSnapshotResponse{Schema: 1, State: "empty", Dataplane: "not-checked", ObservedAt: now.UTC(), MaxAgeSeconds: int64(panelSnapshotLifetime.Seconds()), Admission: a.Operations.Snapshot()}
	if saved := a.panelSnapshots.latest.Load(); saved != nil {
		value.InstanceID, value.Revision = saved.InstanceID, saved.Revision
		generated := saved.Generated
		value.GeneratedAt = &generated
		age := now.Sub(saved.Generated)
		value.DataAgeMS = max(0, age.Milliseconds())
		value.State = "available"
		if age >= 10*time.Second || value.Admission.Exclusive || value.Admission.Fenced {
			value.State = "retained"
		}
		if age >= panelSnapshotLifetime || age < -time.Second {
			value.State = "expired"
		} else {
			value.Data = saved.Data
		}
	}
	// Progress is a separate observation, not part of the configuration revision.
	// Only allowlisted identifiers/counters are exposed, never raw errors, URLs,
	// profile bodies or keys. Metrics are the sampler's already collected value.
	a.nodeChecks.mu.Lock()
	if job := a.nodeChecks.job; job != nil {
		value.Job = &panelSavedJob{job.ID, job.Mode, job.State, job.Phase, job.Total, job.Completed}
	}
	a.nodeChecks.mu.Unlock()
	value.Metrics = a.Stats.Latest()
	return value
}

func (a *App) panelSnapshot(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Cache-Control", "no-store")
	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	w.Header().Set("X-Content-Type-Options", "nosniff")
	if r.Method != http.MethodGet && r.Method != http.MethodHead {
		w.Header().Set("Allow", "GET, HEAD")
		w.WriteHeader(http.StatusMethodNotAllowed)
		return
	}
	if r.Method == http.MethodHead {
		w.WriteHeader(http.StatusOK)
		return
	}
	writeJSON(w, http.StatusOK, a.panelSnapshotAt(time.Now()))
}
